File: TrickyTropolis/Assets/Code/GameUnit.cs
using UnityEngine;
using System.Collections.Generic;

public interface IUnitInfo {
	int getEnergy();
	string getName();
	BuildingCard getOriginCard();
	int? getAttrValue( string attr );
	IEnumerable<string> getAttributes();
}

public class GameUnit : IUnitInfo {
	public string name;
	public int energy;
	public BuildingCard originCard;
	Dictionary<string,int> attributes = new Dictionary<string,int>();
	
	public int getEnergy() { return energy; }
	public string getName() { return name; }
	public BuildingCard getOriginCard() { return originCard; }
	public int? getAttrValue( string attr ) {
		int val;
		if (attributes.TryGetValue (attr, out val))
			return val;
		return null;
	}
	public IEnumerable<string> getAttributes() { 
		return attributes.Keys;
	}

	public void addAttribute( string attr, int val ) {
		if (attributes.ContainsKey (attr)) {
			int currentVal;
			attributes.TryGetValue (attr, out currentVal);
			attributes.Remove( attr );
			attributes.Add (attr, currentVal + val);
		} else {
			attributes.Add (attr, val);
		}
	}
}

File: TrickyTropolis/Assets/Code/Old Java/GameUnit.java
package sequence;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

interface UnitInfo {
	int getEnergy();
	String getName();
	BuildingCard originCard();
	Integer getAttrValue( String attr );
	Set<String> attrSet();
}

public class GameUnit implements UnitInfo {
	String name;
	int energy;
	BuildingCard originCard;
	private Map<String, Integer> attributes = new HashMap<String, Integer>();
	
	public int getEnergy() { return energy; }
	public String getName() { return name; }
	public BuildingCard originCard() { return originCard; }
	public Integer getAttrValue( String attr ) {
		return attributes.get(attr);
	}
	public Set<String> attrSet() { 
		return attributes.keySet();
	}
	public void addAttribute( String attr, int amount ) {
		if( !attributes.containsKey(attr) ) 
			attributes.put(attr, amount);
		else attributes.put( attr, attributes.get(attr) + amount );
	}
}
